File: tools/photo_to_pdf.py
#!/usr/bin/env python3
"""
Turn photographs (or scans) of a document into an uploadable A4 PDF.

For when you printed something, signed it by hand and photographed it: the portal
wants a PDF, not a JPG, and a photo dropped into a PDF unprocessed weighs 8 MB,
comes out crooked and keeps the paper grey.

What it does, in order:
  1. rotates by the EXIF orientation — phone photos are almost always rotated in
     the metadata rather than in the pixels, so some viewers turn them and others
     do not, and your document arrives lying on its side;
  2. converts to greyscale — a document has no colour worth keeping, and grey cuts
     the weight by about two thirds;
  3. flattens the white of the paper, so the background does not stay beige;
  4. resamples to the chosen resolution (200 dpi is plenty for printed text);
  5. lays it out on A4 with a margin, one photo per page;
  6. compresses to JPEG, stepping the resolution down until it fits the size cap.

For heavier correction — uneven lighting, perspective — use clean_scan.py instead.
This one is the straight conversion.

Usage:
    python3 photo_to_pdf.py photo.jpg -o signed.pdf
    python3 photo_to_pdf.py page1.jpg page2.jpg -o signed.pdf --max-mb 1.5
    python3 photo_to_pdf.py scan.png --colour
"""

import argparse
import io
from pathlib import Path

import pymupdf
from PIL import Image, ImageOps

A4 = (595.276, 841.890)      # typographic points, 1 pt = 1/72 inch
MARGIN = 18.0


def prepare(path: Path, dpi: int, colour: bool, flatten_white: bool) -> bytes:
    """Open, straighten and normalise the image. Returns a JPEG in memory."""
    img = Image.open(path)
    img = ImageOps.exif_transpose(img)          # metadata rotation -> pixels

    if not colour:
        img = img.convert("L")
        if flatten_white:
            # Autocontrast with the tails clipped: takes the paper to full white and
            # the ink to black, which is what a scanner app does. The cutoff is low
            # because clipping too hard eats thin pencil strokes and faint signatures.
            img = ImageOps.autocontrast(img, cutoff=(1, 6))
    else:
        img = img.convert("RGB")

    # Usable width on the A4 page in inches, to size the image for that resolution.
    inches = (A4[0] - 2 * MARGIN) / 72
    width_px = int(inches * dpi)
    if img.width > width_px:
        height = round(img.height * width_px / img.width)
        img = img.resize((width_px, height), Image.LANCZOS)

    buf = io.BytesIO()
    img.save(buf, format="JPEG", quality=82, optimize=True)
    return buf.getvalue()


def build(images, out: Path, dpi: int, colour: bool, flatten_white: bool):
    doc = pymupdf.open()
    for path in images:
        data = prepare(path, dpi, colour, flatten_white)
        page = doc.new_page(width=A4[0], height=A4[1])
        frame = pymupdf.Rect(MARGIN, MARGIN, A4[0] - MARGIN, A4[1] - MARGIN)
        # keep_proportion avoids squashing: a squashed page is immediately visible
        # and makes the document look manipulated.
        page.insert_image(frame, stream=data, keep_proportion=True)
    doc.save(str(out), garbage=4, deflate=True)
    doc.close()


def main():
    p = argparse.ArgumentParser(
        description="Photo or scan of a signed document -> A4 PDF.")
    p.add_argument("images", nargs="+", help="one image per page, in order")
    p.add_argument("-o", "--output", default="document.pdf")
    p.add_argument("--dpi", type=int, default=200,
                   help="200 is enough for printed text; 300 if the signature is thin")
    p.add_argument("--colour", "--color", dest="colour", action="store_true",
                   help="keep the colour (only needed for blue ink you want visible)")
    p.add_argument("--no-flatten", action="store_true",
                   help="do not flatten the background (use if the paper comes out washed out)")
    p.add_argument("--max-mb", type=float, default=2.0,
                   help="size cap: below this, upload portals do not complain")
    a = p.parse_args()

    paths = [Path(x) for x in a.images]
    for x in paths:
        if not x.is_file():
            p.error(f"file not found: {x}")

    out = Path(a.output)
    dpi = a.dpi
    # If it overshoots the cap, retry at a lower resolution rather than hand over a
    # file the portal will reject. 120 dpi is the floor: below that, printed text
    # starts to break up.
    while True:
        build(paths, out, dpi, a.colour, not a.no_flatten)
        mb = out.stat().st_size / 1_048_576
        if mb <= a.max_mb or dpi <= 120:
            break
        dpi = max(120, int(dpi * 0.8))
        print(f"  {mb:.1f} MB over the cap, retrying at {dpi} dpi")

    doc = pymupdf.open(out)
    print(f"Wrote: {out}  ({mb:.2f} MB, {len(doc)} pages, {dpi} dpi)")
    if mb > a.max_mb:
        print(f"  WARNING: {mb:.2f} MB, over the {a.max_mb} MB cap. "
              "Retake the photo closer, framing the sheet alone.")
    print("Check at full screen that the signature is clearly visible before uploading.")


if __name__ == "__main__":
    main()

File: docs/app/photo-to-pdf.js
/**
 * The photo-to-PDF page: one photo per A4 page, under a size cap.
 *
 * Straight conversion, in the order tools/photo_to_pdf.py does it: EXIF rotation,
 * greyscale, flatten the paper, resample for the chosen resolution, lay out on A4,
 * and step the resolution down if the file overshoots the cap.
 */

import * as PDFLib from "../vendor/pdf-lib.esm.min.js";
import { toGrey, autocontrastLut } from "../imaging.js";
import { imagesToPdf, pixelWidthForDpi } from "../pdfout.js";
import {
  $, dropZone, status, yieldToPaint, loadImage, pixelsOf, canvasOf,
  resizeCanvas, blobOf, bytesOf, offerDownload, humanSize, escapeHtml,
} from "./ui.js";

const state = $("state");
const listing = $("listing");
const save = $("save");
const notes = $("notes");
let files = [];

dropZone($("drop"), $("picker"), added => {
  files = files.concat(added);
  showList();
  build();
});
for (const id of ["colour", "flatten", "dpi", "cap"]) {
  $(id).addEventListener("change", () => { if (files.length) build(); });
}
$("clear").addEventListener("click", () => {
  files = [];
  showList();
  save.hidden = true;
  notes.innerHTML = "";
  status(state, "");
});

function showList() {
  listing.innerHTML = files.length
    ? "<ol>" + files.map(f => `<li>${escapeHtml(f.name)}</li>`).join("") + "</ol>"
    : "";
  $("clear").hidden = !files.length;
}

/** One image, ready to go on a page: rotated, greyed, flattened, resampled. */
async function prepare(file, dpi) {
  const image = await loadImage(file);
  const px = pixelsOf(image, 4000);
  let canvas;

  if ($("colour").checked) {
    canvas = canvasOf(px.data, px.width, px.height);
  } else {
    const grey = toGrey(px.data, px.width, px.height);
    // Autocontrast with the tails clipped takes the paper to white and the ink to
    // black, which is what a scanner app does. The low cutoff is deliberate: clipping
    // harder eats thin pencil strokes and faint signatures.
    const lut = $("flatten").checked ? autocontrastLut(grey, 1, 6) : null;
    const rgba = new Uint8ClampedArray(px.width * px.height * 4);
    for (let i = 0, p = 0; i < grey.length; i++, p += 4) {
      const v = lut ? lut[Math.round(grey[i])] : grey[i];
      rgba[p] = rgba[p + 1] = rgba[p + 2] = v;
      rgba[p + 3] = 255;
    }
    canvas = canvasOf(rgba, px.width, px.height);
  }

  canvas = resizeCanvas(canvas, pixelWidthForDpi(dpi));
  return bytesOf(await blobOf(canvas, "image/jpeg", 0.82));
}

async function build() {
  save.hidden = true;
  const cap = Number($("cap").value);
  let dpi = Number($("dpi").value);
  const remarks = [];

  try {
    let blob;
    // Overshooting the cap means retrying lower, not handing over a file the upload
    // form will reject. 120 dpi is the floor: below it printed text breaks up.
    for (;;) {
      status(state, `Converting ${files.length} image${files.length > 1 ? "s" : ""} at ${dpi} dpi…`);
      await yieldToPaint();
      const pages = [];
      for (const file of files) pages.push({ bytes: await prepare(file, dpi) });
      blob = new Blob([await imagesToPdf(pages, PDFLib)], { type: "application/pdf" });
      if (blob.size / 1048576 <= cap || dpi <= 120) break;
      remarks.push(`${humanSize(blob.size)} was over the cap, retried lower`);
      dpi = Math.max(120, Math.round(dpi * 0.8));
    }

    offerDownload(save, blob, "document.pdf");
    save.textContent = `Save the PDF (${humanSize(blob.size)})`;
    const mb = blob.size / 1048576;
    if (mb > cap) {
      status(state, `${humanSize(blob.size)}, still over your ${cap} MB cap even at 120 dpi. ` +
                    "Retake the photo closer, with the sheet filling the frame.", "bad");
    } else {
      status(state, `${files.length} page${files.length > 1 ? "s" : ""}, ${humanSize(blob.size)}, ` +
                    `${dpi} dpi. Check at full screen that the signature is legible before ` +
                    "you upload it.", "good");
    }
    notes.innerHTML = remarks.length
      ? "<ul>" + remarks.map(r => `<li>${escapeHtml(r)}</li>`).join("") + "</ul>" : "";
  } catch (err) {
    status(state, `Could not convert that: ${err && err.message || err}`, "bad");
  }
}
